Remove LedgerES dependency from exports

emord opened the pull request:
##### SUMMARY
This introduces a Django model for section and entry combinations in a domain. This idea comes from this conversation https://github.com/dimagi/commcare-hq/pull/25153. I think this is the best option so far in terms of use of system resources and developer time.

This is the last "real" reference to the Ledger ES index now. The remaining ones are in "cleanup" code that will be removed when we remove the index and a [management command](https://github.com/dimagi/commcare-hq/blob/6fb22b04803270979af3692548f5498908d278e4/corehq/apps/dump_reload/management/commands/dump_domain_data_raw.py) that may or may not have been used since 2017. @snopoke looks like you introduced that for a particular project. Is it still necessary, or can we remove it? Either way it doesn't look crazy hard to rewrite to iterate over SQL backend.

I didn't remove the actual ES index here because keeping it in the short term allows us to revert if necessary. I'll plan to remove it when I come back from PTO in a couple weeks.

For me to do tomorrow: 

- [x] Create a script to initialize the data in `LedgerSectionEntry`

emord committed:
Create model for section and entry combinations

emord committed:
Update a domain's ledger combinations in the ledger pillow

emord committed:
Replace ES query with postgres

snopoke commented:
> This is the last "real" reference to the Ledger ES index now. The remaining ones are in "cleanup" code that will be removed when we remove the index and a management command that may or may not have been used since 2017. @snopoke looks like you introduced that for a particular project. Is it still necessary, or can we remove it? Either way it doesn't look crazy hard to rewrite to iterate over SQL backend.

It's only been used a couple of times. Fine to remove the ledgers support. We can add it back later if it's needed.

snopoke commented on corehq/pillows/ledger.py:
```diff
@@ -57,6 +62,27 @@ def _get_daily_consumption_for_ledger(ledger):
+        return
+
+    # clear the lru_cache so that next time a ledger is saved, we get the combinations
+    _get_ledger_section_combinations.cache_clear()
```
shouldn't this happen after the new entry is added

emord replied:
Technically since I'm using a local LRU cache then I think this was safe before, but clearing cache later does make sense if there was a chance for parallel queries https://github.com/dimagi/commcare-hq/pull/25304/commits/f9184cf162e4005d5ca3ff975bbac0250892b994

snopoke commented on corehq/apps/export/models/new.py:
```diff
@@ -2772,6 +2769,15 @@ class EmailExportWhenDoneRequest(models.Model):
     user_id = models.CharField(max_length=255)
 
 
+class LedgerSectionEntry(models.Model):
```
what do you think about bootstrapping this table with data from ES?

emord replied:
Yeah. That's what I meant by the TODO in the description. implemnted here: https://github.com/dimagi/commcare-hq/pull/25304/commits/f9184cf162e4005d5ca3ff975bbac0250892b994

snopoke commented on corehq/pillows/ledger.py:
```diff
@@ -37,6 +40,8 @@ def _prepare_ledger_for_es(ledger):
     if not ledger.get('location_id') and ledger.get('case_id'):
         ledger['location_id'] = _location_id_for_case(ledger['case_id'])
 
+    _update_ledger_section_entry_combinations(ledger)
```
where will this get called when the pillow get's removed?

emord replied:
So my original intention was to totally remove the pillow, but then I found that this is currently how we do daily consumption after running the full test suite. 

My current plan is to keep the pillow, but remove the ES index as that seemed like a good use of the work I had already done.

So in other words, this pillow will stay and it will contain consumption updating and this update for now

emord committed:
Correct usage of lru_cache

emord committed:
Reset cache after saving new record

emord committed:
Add initialization for LedgerSectionEntry

snopoke commented on corehq/apps/export/migrations/0008_auto_20190906_2008.py:
```diff
@@ -0,0 +1,40 @@
+    ]
+
+    operations = [
+        migrations.RunPython(initialize_ledger_combinations),
```
should make this elidable with `elidable=True`

emord replied:
```suggestion
        migrations.RunPython(initialize_ledger_combinations, elidable=True),
```

snopoke reviewed (approved):
One comment but not blocking

emord committed:
Update corehq/apps/export/migrations/0008_auto_20190906_2008.py

emord committed:
Skip data migration on fresh install

emord committed:
Fix mock for stock export column

The function now guarantees that the columns are sorted from the
database.

The unique index also ensures there are no duplicates

emord commented:
The migration has been applied to staging, prod and TCL in a private release. I also ensure the tables looked fine after the migration completed.
